// @sanjilraj/pern-auth-utils/src/index.js
const jwt = require('jsonwebtoken');

// ─── Token Creation ───────────────────────────────────────────────────────────

/**
 * Creates a signed JWT token for a user.
 * @param {object} user - User object with user_id, role, email
 * @param {string} secret - JWT secret key
 * @param {string} expiresIn - Token expiry (default: '7d')
 * @returns {string} Signed JWT token
 */
const createToken = (user, secret, expiresIn = '7d') => {
    if (!secret) throw new Error('JWT_SECRET is required');
    return jwt.sign(
        { userId: user.user_id, role: user.role, email: user.email },
        secret,
        { expiresIn }
    );
};

// ─── Token Verification ───────────────────────────────────────────────────────

/**
 * Verifies a JWT token and returns the decoded payload.
 * @param {string} token - JWT token string
 * @param {string} secret - JWT secret key
 * @returns {object} Decoded token payload
 */
const verifyToken = (token, secret) => {
    if (!secret) throw new Error('JWT_SECRET is required');
    return jwt.verify(token, secret);
};

// ─── Middleware Factory ───────────────────────────────────────────────────────

/**
 * Creates an Express middleware that authenticates JWT tokens.
 * Reads token from Authorization: Bearer <token> header.
 * @param {string} secret - JWT secret key
 * @returns {Function} Express middleware
 */
const createAuthMiddleware = (secret) => (req, res, next) => {
    const authHeader = req.headers.authorization || '';
    const token = authHeader.startsWith('Bearer ') ? authHeader.slice(7) : null;
    if (!token) {
        return res.status(401).json({ error: 'Missing authentication token' });
    }
    try {
        req.auth = jwt.verify(token, secret);
        return next();
    } catch {
        return res.status(401).json({ error: 'Invalid or expired authentication token' });
    }
};

// ─── Admin Guard Middleware ───────────────────────────────────────────────────

/**
 * Express middleware that restricts access to admin users only.
 * Must be used AFTER loadCurrentUser middleware (req.currentUser must exist).
 */
const requireAdmin = (req, res, next) => {
    if (!req.currentUser) {
        return res.status(401).json({ error: 'Authentication required' });
    }
    if (req.currentUser.role !== 'admin') {
        return res.status(403).json({ error: 'Admin access required' });
    }
    return next();
};

// ─── User Access Guard ────────────────────────────────────────────────────────

/**
 * Checks if the current user can access a target user's resources.
 * Admins can access any user; regular users can only access their own.
 * @param {object} req - Express request object (req.currentUser must exist)
 * @param {number} targetUserId - The user ID being accessed
 * @returns {boolean}
 */
const canAccessUser = (req, targetUserId) => {
    if (!req.currentUser) return false;
    return req.currentUser.role === 'admin' ||
        req.currentUser.user_id === Number(targetUserId);
};

// ─── User Sanitizer ───────────────────────────────────────────────────────────

/**
 * Removes sensitive fields from a user object before sending to client.
 * Strips password_hash, raw provider_id, etc.
 * @param {object} user - Raw user object from database
 * @returns {object} Safe user object for API responses
 */
const sanitizeUser = (user) => ({
    id: user.user_id,
    name: user.name,
    email: user.email,
    role: user.role,
    provider: user.provider || null,
    avatarUrl: user.avatar_url || null,
    createdAt: user.created_at,
    updatedAt: user.updated_at,
});

// ─── OAuth State Helpers ──────────────────────────────────────────────────────

const crypto = require('crypto');

/**
 * Generates a cryptographically secure random OAuth state string.
 * @returns {string} Random hex string
 */
const createOAuthState = () => crypto.randomBytes(24).toString('hex');

/**
 * Returns the cookie name used to store OAuth state for a provider.
 * @param {string} provider - OAuth provider name (e.g. 'google', 'facebook')
 * @returns {string}
 */
const getOAuthStateCookieName = (provider) => `oauth_state_${provider}`;

/**
 * Sets the OAuth state cookie on the response.
 * @param {object} res - Express response
 * @param {string} provider - OAuth provider name
 * @param {string} state - State value to store
 * @param {object} cookieOptions - Cookie options
 */
const setOAuthStateCookie = (res, provider, state, cookieOptions = {}) => {
    const defaults = {
        httpOnly: true,
        sameSite: 'lax',
        secure: process.env.NODE_ENV === 'production',
        path: '/',
    };
    res.cookie(getOAuthStateCookieName(provider), state, { ...defaults, ...cookieOptions });
};

/**
 * Verifies the OAuth state cookie matches the query param state.
 * Clears the state cookie after verification.
 * @param {object} req - Express request
 * @param {object} res - Express response
 * @param {string} provider - OAuth provider name
 * @returns {boolean} True if state matches, false otherwise
 */
const verifyOAuthState = (req, res, provider) => {
    const cookieName = getOAuthStateCookieName(provider);
    const cookieState = req.cookies[cookieName];
    const queryState = req.query.state;
    res.clearCookie(cookieName, { path: '/' });
    if (!cookieState || !queryState || cookieState !== queryState) return false;
    return true;
};

module.exports = {
    createToken,
    verifyToken,
    createAuthMiddleware,
    requireAdmin,
    canAccessUser,
    sanitizeUser,
    createOAuthState,
    getOAuthStateCookieName,
    setOAuthStateCookie,
    verifyOAuthState,
};
